### PR TITLE
Implement narrow_copy derivative

### LINKED ISSUE #65617: Derivative not implemented for narrow_copy
```
>>> torch.narrow_copy(torch.zeros(2, requires_grad=True), 0, 0, 1)
tensor([0.], grad_fn=<NotImplemented>)
```

cc @bdhirsh

### PR DESCRIPTION
**Pull request by cluePrints**
Fixes #65617

<details><summary>Unit testing ok</summary>
### Test ok ✅
```
 python test/test_torch.py -k "test_narrow_copy_gradient" -v
test_narrow_copy_gradient_cpu (__main__.TestTorchDeviceTypeCPU) ... ok
...
Ran 1 test in 0.022s
OK
```

### Linter ok ✅
```
 lintrunner f 
ok No lint issues.
```

### Existing tests ok ✅
```
pip install -e . -v --no-build-isolation && python test/test_ops.py -k "narrow_copy" -v && python test/test_ops_gradients.py -k "narrow_copy" -v 
```
https://www.diffchecker.com/JIeLTegz/
```
python test/test_mps.py TestConsistencyMPS.test_output_grad_match_narrow_copy_mps_float16 TestConsistencyMPS.test_output_grad_match_narrow_copy_mps_float32
..
----------------------------------------------------------------------
Ran 2 tests in 0.143s

OK
```
</details>

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

**Review by soulitzer — APPROVED** [review]

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -18077,9 +18077,8 @@ def sample_inputs_abs(op_info, device, dtype, requires_grad, op_kwargs=None, **k
     OpInfo('narrow_copy',
            dtypes=all_types_and_complex_and(torch.bool, torch.bfloat16, torch.float16, torch.chalf),
            supports_out=True,
-           supports_forward_ad=False,
-           supports_fwgrad_bwgrad=False,
-           supports_autograd=False,
+           supports_forward_ad=True,
+           supports_fwgrad_bwgrad=True,
            # https://github.com/pytorch/pytorch/issues/86931
            sample_inputs_func=partial(sample_inputs_narrow_narrow_copy, is_narrow=False),
```
Comment (Skylion007) [inline torch/testing/_internal/common_methods_invocations.py:18078]: Doesn't sample input functions cover the edge cases added in the test cases?

**Comment by Skylion007** [comment]
You have to fix some unexpected successes on MPS test suite. Please remove the relevant xfails from there.

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (cluePrints) [reply torch/testing/_internal/common_methods_invocations.py:18078]: @Skylion007  Oh, I didn't realize the meta-test here checks the gradient vals – that was my primary goal
https://github.com/pytorch/pytorch/blob/3258b1853508fa63ecfa9faa9192f7db1a16c1d8/torch/testing/_internal/common_utils.py#L5781

Had a look at the cases, I think the only extra case is dim=0 but I don't think it's adding much value so I'm gonna drop extra stuff in test_narrow_copy_gradient
https://github.com/pytorch/pytorch/blob/3258b1853508fa63ecfa9faa9192f7db1a16c1d8/torch/testing/_internal/common_methods_invocations.py#L5672-L5679

**Review by cluePrints — COMMENTED** [review]
Dropped unnecessary test (test_narrow_copy_gradient) and enabled a one which now passes (TestConsistencyMPS.test_output_grad_match_narrow_copy_mps_float{16,32}). Also rebased against main
